Feature: AES Encryption and Decryption

Jordan10980 opened the pull request:
This pull request is a work in progress for adding AES encryption and decryption functionality. It includes the implementation of file encryption and decryption using AES, as well as key generation and management for AES256 encryption. The changes also come with a comprehensive set of tests to ensure the reliability and security of the new features.

Jordan10980 committed:
feat(nodeid):Generate id for the futur nodes

Jordan10980 committed:
Merge branch 'main' of https://github.com/pierreleocadie/SecuraChain

Jordan10980 committed:
feat(aeskey): Implement AES encryption and decryption
Implemented functions to encrypt and decrypt files using AES encryption.
Added support to generate and manage 32-byte keys for AES256 encryption and decryption.
Conducted thorough testing for AES key generation, encryption, and decryption.

pierreleocadie commented on pkg/nodes_functions/idnode.go:
The package "nodes_functions" must be deleted.

Jordan10980 replied:
The package "nodes_functions" doesn't exists anymore

pierreleocadie replied:
https://github.com/pierreleocadie/SecuraChain/tree/aeskey/pkg/nodes_functions

Don't forget to commit and push.

pierreleocadie commented on pkg/aeskey/aesKeyGen.go:
- The KeySize constant doesn't need to be exported, so you should keep it unexported by renaming it keySize instead.
- You have to choose between leaving your structure as it is (exported), in which case you should export its "key" field too so that you no longer need a getter, or making the structure private and its field private, then using an exported interface that you can name Key (here you have to return the Key type with the NewKey), for example with the structure's exported methods, which are String() and Key(). Here you will need a getter to access the "key" field, but you should name the getter "Key" and not "GetKey" (see Effective Go documentation link below on getters guidelines). You can discuss this with our mutual best friend to help you decide on this encapsulation.

cf. : https://go.dev/doc/effective_go#Getters

pierreleocadie commented on pkg/aeskey/cryptFile.go:
Look what I've done to clean up the path to prevent any potential attempts to traverse the path: https://github.com/pierreleocadie/SecuraChain/blob/e4fd9e6f01e3df0071adfd4e5f269a0fb64a9172/pkg/ecdsa/storage.go#L83-L91

I think you should do the same thing, returning an error if you detect an attempt.
The EncryptFile and DecryptFile functions must be structure functions, which avoids passing the key as a parameter to the function.

pierreleocadie commented:
You should rename the package to aes instead of aes key to be consistent with the renaming of the ecdsakeypair package to ecdsa.

https://go.dev/doc/effective_go#Getters

pierreleocadie reviewed (changes requested):
Take my comments into consideration and rename the package to aes instead of aes key to be consistent with the renaming of the ecdsakeypair package to ecdsa.

Don't forget to update your branch with the main branch latest changes.

(PS: All checks must be passed and validated to merge onto main)

https://go.dev/doc/effective_go#Getters

pierreleocadie committed:
Merge branch 'main' into aeskey

pierreleocadie committed:
refactor: update aes package, add documentation, externalize code, and modify tests

This commit includes several changes to the aes package and related code. First, the pull request from Jordan for merging the aeskey branch into main has been updated to address requested changes. Second, the aes package has been refactored for improved readability and maintainability. Third, documentation has been added to the aes package for improved usability. Fourth, code for sanitizing paths and loading PEM files has been externalized to a new utils package in the pkg directory. Finally, the test for the ecdsa package has been modified to remove generated PEM files.

pierreleocadie reviewed (approved):
I fixed everything. It can be merged rn.
